core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPostProcessing.kt

```diff
@@ -120,106 +122,142 @@ class STDCMPostProcessing(private val graph: STDCMGraph) {
-    val res = ArrayList<PathfindingEdgeLocationId<Block>>()
-    for (waypoint in path.waypoints) {
-        val blockOffset = waypoint.edge.blockOffsetFromEdge(waypoint.offset)
-        res.add(EdgeLocation(waypoint.edge.block, blockOffset))
+    /** Creates the list of waypoints on the path */
+    private fun makeBlockWaypoints(path: Result): List<PathfindingEdgeLocationId<Block>> {
+        val res = ArrayList<PathfindingEdgeLocationId<Block>>()
+        for (waypoint in path.waypoints) {
+            val blockOffset = waypoint.edge.blockOffsetFromEdge(waypoint.offset)
+            res.add(EdgeLocation(waypoint.edge.block, blockOffset))
+        }
+        return res
     }
-    return res
-}
 
-/** Computes the departure time, made of the sum of all delays added over the path */
-fun computeDepartureTime(edges: List<STDCMEdge>, startTime: Double): Double {
-    var mutStartTime = startTime
-    for (edge in edges) mutStartTime += edge.addedDelay
-    return mutStartTime
-}
+    /** Computes the departure time, made of the sum of all delays added over the path */
+    private fun computeDepartureTime(edges: List<STDCMEdge>, startTime: Double): Double {
+        var addedDelay = 0.0
+        for (edge in edges) addedDelay += edge.addedDelay
+        val totalAddedDelay = addDelayForPlannedNodes(edges, addedDelay)
+        return startTime + totalAddedDelay
+    }
 
-/** Builds the list of stops from the edges */
-private fun makeStops(edges: List<STDCMEdge>): List<TrainStop> {
-    val res = ArrayList<TrainStop>()
-    var offset = 0.meters
-    for (edge in edges) {
-        val prevNode = edge.previousNode
-        // Ignore first path node and last node (we aren't checking lastEdge.getEdgeEnd())
-        if (
-            prevNode.previousEdge != null &&
-                prevNode.stopDuration != null &&
-                prevNode.stopDuration >= 0
-        )
-            res.add(
-                TrainStop(
-                    offset.meters,
-                    prevNode.stopDuration,
-                    // TODO: forward and use onStopSignal param from request
-                    isTimeStrictlyPositive(prevNode.stopDuration)
-                )
+    /** Add delay needed to minimize the relative time diff with planned timing data */
+    private fun addDelayForPlannedNodes(edges: List<STDCMEdge>, addedDelay: Double): Double {
+        val nodes = getNodes(edges)
+        val maxNodeDelay = getMaxNodeDelay(nodes)
+        val plannedNodes = getPlannedNodes(nodes)
+        var delayForPlannedNodes = 0.0
+        for (plannedNode in plannedNodes) {
+            val timeDiff = plannedNode.getTimeDiffWithPlannedArrival(addedDelay)
+            // Train passes before planned node: adding delay can make the train pass closer to
+            // planned arrival time
+            if (timeDiff < 0.0) {
+                delayForPlannedNodes = abs(timeDiff)
+                break
+            }
+        }
```

Comment:
Do we really need a loop? We can just pick any value here

If `timeDiff >= 0.0` for any node we can just not do anything (adding delay would add difference for at least one node)

Reply:
Would work as well. Just to be sure, you're proposing the following: we take the first planned node, if `timeDiff >= 0` we add nothing (even if further along the way we arrive before a planned node's arrival time)?
I'll also add a corresponding unit test if so.